Illegal instruction. Android

I did everything as in BUILD_Andorid.md And use drivers ```mesa-opencl-icd-rusticl/stable,now 25.3.0-1 aarch64```

![Screenshot_2025-12-02-01-32-50-889_com.termux.jpg](https://github.com/user-attachments/assets/922194f7-724c-4685-a7b3-46189dc180a6)

> > [@ahmed-alnassif](https://github.com/ahmed-alnassif)
> > [build.zip](https://github.com/user-attachments/files/24941096/build.zip)
> 
> After reviewing the log there are no errors and the build is done successfully, and I think the issue is because you're using `clang` instead of `gcc` and `clang` can't detect CPU features in some Android kernels (Qualcomm Snapdragon in your case).
> 
> You can try setting them and build hashcat:
> 
> make clean
> make CFLAGS="-O2 -march=armv8-a -mtune=generic"
> There is no issue in hashcat source code and this issue is device-specific.

```
~/hashcat $ make CFLAGS="-O2 -march=armv8-a -mtune=generic"
Android environment detected.
gcc -c -std=gnu99 -O2 -march=armv8-a -mtune=generic -D__ANDROID__ src/affinity.c -o obj/affinity.NATIVE.o -fpic
gcc -c -std=gnu99 -O2 -march=armv8-a -mtune=generic -D__ANDROID__ src/autotune.c -o obj/autotune.NATIVE.o -fpic
gcc -c -std=gnu99 -O2 -march=armv8-a -mtune=generic -D__ANDROID__ src/backend.c -o obj/backend.NATIVE.o -fpic
gcc -c -std=gnu99 -O2 -march=armv8-a -mtune=generic -D__ANDROID__ src/benchmark.c -o obj/benchmark.NATIVE.o -fpic
gcc -c -std=gnu99 -O2 -march=armv8-a -mtune=generic -D__ANDROID__ src/bitmap.c -o obj/bitmap.NATIVE.o -fpic
gcc -c -std=gnu99 -O2 -march=armv8-a -mtune=generic -D__ANDROID__ src/bitops.c -o obj/bitops.NATIVE.o -fpic
gcc -c -std=gnu99 -O2 -march=armv8-a -mtune=generic -D__ANDROID__ src/bridges.c -o obj/bridges.NATIVE.o -fpic
gcc -c -std=gnu99 -O2 -march=armv8-a -mtune=generic -D__ANDROID__ src/combinator.c -o obj/combinator.NATIVE.o -fpic
src/benchmark.csrc/bitmap.csrc/affinity.csrc/bridges.csrc/combinator.c::::6:src/bitops.c66::106:610:::10::10 :6: 10:  :10 fatal error: fatal error: :fatal error: fatal error:  'common.h'fatal error:  'common.h'file'common.h''common.h''common.h'    filefatal error:  filenotfile
 file        notfound
       notnot
notfound


found
'common.h'
foundfound

file        6 6 |     #6i |     not     | n
#66 | #ci       | #iln#foundinucincdlnclleu
cuu ddlde" euec d o"emc" "moccomo"monmcmm    .ooommhn6nmo"..o | nhhn.
#"".h      ih"|

"n      c
         ^~~~~~~~~~| |
               ^~~~~~~~~~| l
u
|          ^~~~~~~~~~
         ^~~~~~~~~~d
         ^~~~~~~~~~e
 "commosrc/autotune.cn:.6h:"10:
       |          ^~~~~~~~~~
fatal error: src/backend.c:6'common.h': 10file:  not
      fatal error: found
'common.h' file not
          6found | #
include "common.h"    6
 |       #| in         ^~~~~~~~~~c
lude "common.h"
      |          ^~~~~~~~~~
1 error generated.
1 error generated.
make: *** [src/Makefile:850: obj/bitops.NATIVE.o] Error 1
make: *** Waiting for unfinished jobs....
make: *** [src/Makefile:850: obj/benchmark.NATIVE.o] Error 1
1 error generated.
11 error error generated generated.
.
1 error generated.
1 error generated.
make: *** [src/Makefile:850: obj/combinator.NATIVE.o] Error 1
make: *** [src/Makefile:850: obj/bridges.NATIVE.o] Error 1
make: *** [src/Makefile:850: obj/bitmap.NATIVE.o] Error 1
make: *** [src/Makefile:850: obj/affinity.NATIVE.o] Error 1
make: *** [src/Makefile:850: obj/autotune.NATIVE.o] Error 1
1 error generated.
make: *** [src/Makefile:850: obj/backend.NATIVE.o] Error 1
~/hashcat $
```

> > > [@ahmed-alnassif](https://github.com/ahmed-alnassif)
> > > [build.zip](https://github.com/user-attachments/files/24941096/build.zip)
> > 
> > After reviewing the log there are no errors and the build is done successfully, and I think the issue is because you're using `clang` instead of `gcc` and `clang` can't detect CPU features in some Android kernels (Qualcomm Snapdragon in your case).
> > 
> > You can try setting them and build hashcat:
> > 
> > make clean
> > make CFLAGS="-O2 -march=armv8-a -mtune=generic"
> > There is no issue in hashcat source code and this issue is device-specific.
> 
> ```
> ~/hashcat $ make CFLAGS="-O2 -march=armv8-a -mtune=generic"
> Android environment detected.
> gcc -c -std=gnu99 -O2 -march=armv8-a -mtune=generic -D__ANDROID__ src/affinity.c -o obj/affinity.NATIVE.o -fpic
> gcc -c -std=gnu99 -O2 -march=armv8-a -mtune=generic -D__ANDROID__ src/autotune.c -o obj/autotune.NATIVE.o -fpic
> gcc -c -std=gnu99 -O2 -march=armv8-a -mtune=generic -D__ANDROID__ src/backend.c -o obj/backend.NATIVE.o -fpic
> gcc -c -std=gnu99 -O2 -march=armv8-a -mtune=generic -D__ANDROID__ src/benchmark.c -o obj/benchmark.NATIVE.o -fpic
> gcc -c -std=gnu99 -O2 -march=armv8-a -mtune=generic -D__ANDROID__ src/bitmap.c -o obj/bitmap.NATIVE.o -fpic
> gcc -c -std=gnu99 -O2 -march=armv8-a -mtune=generic -D__ANDROID__ src/bitops.c -o obj/bitops.NATIVE.o -fpic
> gcc -c -std=gnu99 -O2 -march=armv8-a -mtune=generic -D__ANDROID__ src/bridges.c -o obj/bridges.NATIVE.o -fpic
> gcc -c -std=gnu99 -O2 -march=armv8-a -mtune=generic -D__ANDROID__ src/combinator.c -o obj/combinator.NATIVE.o -fpic
> src/benchmark.csrc/bitmap.csrc/affinity.csrc/bridges.csrc/combinator.c::::6:src/bitops.c66::106:610:::10::10 :6: 10:  :10 fatal error: fatal error: :fatal error: fatal error:  'common.h'fatal error:  'common.h'file'common.h''common.h''common.h'    filefatal error:  filenotfile
>  file        notfound
>        notnot
> notfound
> 
> 
> found
> 'common.h'
> foundfound
> 
> file        6 6 |     #6i |     not     | n
> #66 | #ci       | #iln#foundinucincdlnclleu
> cuu ddlde" euec d o"emc" "moccomo"monmcmm    .ooommhn6nmo"..o | nhhn.
> #"".h      ih"|
> 
> "n      c
>          ^~~~~~~~~~| |
>                ^~~~~~~~~~| l
> u
> |          ^~~~~~~~~~
>          ^~~~~~~~~~d
>          ^~~~~~~~~~e
>  "commosrc/autotune.cn:.6h:"10:
>        |          ^~~~~~~~~~
> fatal error: src/backend.c:6'common.h': 10file:  not
>       fatal error: found
> 'common.h' file not
>           6found | #
> include "common.h"    6
>  |       #| in         ^~~~~~~~~~c
> lude "common.h"
>       |          ^~~~~~~~~~
> 1 error generated.
> 1 error generated.
> make: *** [src/Makefile:850: obj/bitops.NATIVE.o] Error 1
> make: *** Waiting for unfinished jobs....
> make: *** [src/Makefile:850: obj/benchmark.NATIVE.o] Error 1
> 1 error generated.
> 11 error error generated generated.
> .
> 1 error generated.
> 1 error generated.
> make: *** [src/Makefile:850: obj/combinator.NATIVE.o] Error 1
> make: *** [src/Makefile:850: obj/bridges.NATIVE.o] Error 1
> make: *** [src/Makefile:850: obj/bitmap.NATIVE.o] Error 1
> make: *** [src/Makefile:850: obj/affinity.NATIVE.o] Error 1
> make: *** [src/Makefile:850: obj/autotune.NATIVE.o] Error 1
> 1 error generated.
> make: *** [src/Makefile:850: obj/backend.NATIVE.o] Error 1
> ~/hashcat $
> ```


`CFLAGS` flag have overwrite the Makefile flags, so you will need to edit src/Makefile to set them.

You can use `sed` command to do this:
```bash
sed -i 's/-march=native/-march=armv8-a/g' src/Makefile
sed -i 's/-mtune=native/-mtune=generic/g' src/Makefile

make clean
make
```
Since the issue is device-specific I can't help you further.

--- a/BUILD_Android.md
+++ b/BUILD_Android.md
@@ -6,7 +6,7 @@ Author: See `docs/credits.txt`
 
 ---
 
-## ✅ Android Requirements
+## Android Requirements
 
 - Android 8.0 or higher (OpenCL 3.0 support)
 
@@ -14,7 +14,7 @@ Author: See `docs/credits.txt`
 
 ---
 
-## 🛠️ Building Hashcat on Android
+## Building Hashcat on Android
 
 ### Step 1: Install Dependencies
 
@@ -115,7 +115,7 @@ Verify the fix worked:
 
 ---
 
-## 🔧 OpenCL Troubleshooting
+## OpenCL Troubleshooting
 
 If Library Not Found on Device
 
@@ -151,7 +151,7 @@ clinfo | grep "Number of platforms"
 
 ---
 
-## 🚀 Usage Examples
+## Usage Examples
 
 Safe Benchmark (Skips Memory-Intensive Algorithms)
 
@@ -173,7 +173,7 @@ Brute Force
 
 ---
 
-## ⚠️ Known Limitations
+## Known Limitations
 
 Memory-Intensive Algorithms
 
@@ -199,21 +199,21 @@ Recommended for Mobile
 
 ---
 
-## 🚀 Performance Results (POCO X6 Pro)
+## Performance Results (POCO X6 Pro)
 
-| Algorithm | Speed     | Status |
-|-----------|-----------|--------|
-| MD4       | 1179 MH/s | ✅     |
-| MD5       | 853 MH/s  | ✅     |
-| SHA1      | 282 MH/s  | ✅     |
-| SHA256    | 111 MH/s  | ✅     |
-| WPA2      | OOM       | ❌     |
+| Algorithm | Speed     | Status  |
+|-----------|-----------|---------|
+| MD4       | 1179 MH/s | Working |
+| MD5       | 853 MH/s  | Working |
+| SHA1      | 282 MH/s  | Working |
+| SHA256    | 111 MH/s  | Working |
+| WPA2      | OOM       | Crashed |
 
 Real-world: 9-character password cracked in 90 seconds at 694 MH/s
 
 ---
 
-## 🎉 Done
+## Done
 
 Your Android device is now ready for **hashcat!** Perfect for:
 
@@ -227,4 +227,4 @@ Your Android device is now ready for **hashcat!** Perfect for:
 
 ---
 
-Tested on **POCO X6 Pro** • **Android 16** • **Termux 0.119.0-beta.3**
+Tested on **POCO X6 Pro** | **Android 16** | **Termux 0.119.0-beta.3**

--- a/BUILD_Android.md
+++ b/BUILD_Android.md
@@ -23,7 +23,7 @@ Step 1: Install Dependencies
 ```bash
 apt update
 apt install git make clang python rust libiconv sse2neon
-apt install libbz2 liblzma libsqlite openssl readline ncurses
+apt install libbz2 liblzma libsqlite openssl readline ncurses ncurses-utils
 ```
 
 Install OpenCL Vendor Driver:

--- a/BUILD_Android.md
+++ b/BUILD_Android.md
@@ -22,18 +22,24 @@ Step 1: Install Dependencies
 
 ```bash
 apt update
-apt install git make clang python rust libiconv sse2neon opencl-vendor-driver
+apt install git make clang python rust libiconv sse2neon
 apt install libbz2 liblzma libsqlite openssl readline ncurses
+```
+
+Install OpenCL Vendor Driver:
+
+```bash
+apt install -y opencl-vendor-driver
 ```
 
 Step 2: Verify OpenCL Support
 
 ```bash
 apt install -y clinfo
-clinfo
+clinfo -l
 ```
 
-If clinfo shows 0 platforms, you need to fix OpenCL libraries.
+If clinfo shows nothing, you will need to fix OpenCL libraries.
 
 Step 3: Fix OpenCL Libraries (If Needed)
 
@@ -104,7 +110,7 @@ If the vendor driver doesn't work:
 
 ```bash
 apt remove opencl-vendor-driver
-apt install mesa-opencl-icd
+apt install mesa-opencl-icd-rusticl
 ```
 
 Verify OpenCL Fix

--- a/BUILD_Android.md
+++ b/BUILD_Android.md
@@ -96,6 +96,23 @@ make clean; make
 ./hashcat --version
 ```
 
+> [!IMPORTANT]
+> If you encounter an `Illegal instruction` error, you may need to adjust the target architecture. This is a known issue for some ARM devices ([#4579](https://github.com/hashcat/hashcat/issues/4579)).
+
+```bash
+sed -i 's/-march=native/-march=armv8-a/g' src/Makefile
+sed -i 's/-mtune=native/-mtune=generic/g' src/Makefile
+
+make clean
+make
+```
+
+Verify the fix worked:
+
+```bash
+./hashcat -I
+```
+
 ---
 
 ## 🔧 OpenCL Troubleshooting

--- a/BUILD_Android.md
+++ b/BUILD_Android.md
@@ -6,19 +6,17 @@ Author: See `docs/credits.txt`
 
 ---
 
-✅ Android Requirements
+## ✅ Android Requirements
 
-· Android 8.0 or higher (OpenCL 3.0 support)
+- Android 8.0 or higher (OpenCL 3.0 support)
 
-· Termux app installed from F-Droid or [GitHub](https://github.com/termux/termux-app/releases/latest)
-
-· ARM64 device with OpenCL-capable GPU
+- Termux app installed from [GitHub](https://github.com/termux/termux-app/releases/latest) or F-Droid
 
 ---
 
-🛠️ Building Hashcat on Android
+## 🛠️ Building Hashcat on Android
 
-Step 1: Install Dependencies
+### Step 1: Install Dependencies
 
 ```bash
 apt update
@@ -32,7 +30,7 @@ Install OpenCL Vendor Driver:
 apt install -y opencl-vendor-driver
 ```
 
-Step 2: Verify OpenCL Support
+### Step 2: Verify OpenCL Support
 
 ```bash
 apt install -y clinfo
@@ -41,12 +39,19 @@ clinfo -l
 
 If clinfo shows nothing, you will need to fix OpenCL libraries.
 
-Step 3: Fix OpenCL Libraries (If Needed)
+### Step 3: Fix OpenCL Libraries (If Needed)
 
 Auto-Fix (Try This First):
 
 ```bash
-lib_path=$(find /system /vendor -name "android.hardware.graphics.common-V*-ndk.so" 2>/dev/null | head -1) && if [ -n "$lib_path" ]; then mkdir -p $PREFIX/opt/vendor/lib && ln -sf "$lib_path" "$PREFIX/opt/vendor/lib/android.hardware.graphics.common-V4-ndk.so" && echo "✅ OpenCL linked: $lib_path"; else echo "❌ Auto-fix failed - try manual linking"; fi
+lib_path=$(find /system/lib64 /vendor/lib64 /system/lib /vendor/lib -name "android.hardware.graphics.common-V*-ndk.so" 2>/dev/null | head -1) && \
+if [ -n "$lib_path" ]; then \
+    mkdir -p $PREFIX/opt/vendor/lib && \
+    ln -sf "$lib_path" "$PREFIX/opt/vendor/lib/android.hardware.graphics.common-V4-ndk.so" && \
+    echo "[+] OpenCL linked: $lib_path"; \
+else \
+    echo "[-] Auto-fix failed - try manual linking"; \
+fi
 ```
 
 Manual Solution (If Auto-Fix Failed):
@@ -61,15 +66,22 @@ Link the library:
 
 ```bash
 mkdir -p $PREFIX/opt/vendor/lib
-ln -s /path/to/found/library $PREFIX/opt/vendor/lib/android.hardware.graphics.common-V4-ndk.so
+ln -s /path/to/found/library.so $PREFIX/opt/vendor/lib/android.hardware.graphics.common-V4-ndk.so
 ```
 
-Example:
+Example (64-bit):
 
 ```bash
 ln -s /system/lib64/android.hardware.graphics.common-V5-ndk.so $PREFIX/opt/vendor/lib/android.hardware.graphics.common-V4-ndk.so
 ```
-Step 4: Clone and Build Hashcat
+
+Example (32-bit):
+
+```bash
+ln -s /system/lib/android.hardware.graphics.common-V5-ndk.so $PREFIX/opt/vendor/lib/android.hardware.graphics.common-V4-ndk.so
+```
+
+### Step 4: Clone and Build Hashcat
 
 ```bash
 git clone --depth 1 https://github.com/hashcat/hashcat.git
@@ -77,7 +89,7 @@ cd hashcat
 make clean; make
 ```
 
-Step 5: Verify Build
+### Step 5: Verify Build
 
 ```bash
 ./hashcat -I
@@ -86,7 +98,7 @@ Step 5: Verify Build
 
 ---
 
-🔧 OpenCL Troubleshooting
+## 🔧 OpenCL Troubleshooting
 
 If Library Not Found on Device
 
@@ -122,7 +134,7 @@ clinfo | grep "Number of platforms"
 
 ---
 
-🚀 Usage Examples
+## 🚀 Usage Examples
 
 Safe Benchmark (Skips Memory-Intensive Algorithms)
 
@@ -144,29 +156,29 @@ Brute Force
 
 ---
 
-⚠️ Known Limitations
+## ⚠️ Known Limitations
 
 Memory-Intensive Algorithms
 
 These algorithms exceed mobile memory limits and will crash:
 
-· WPA2 (22000) - PBKDF2 memory requirements
+- WPA2 (22000) - PBKDF2 memory requirements
 
-· Bitcoin (11300) - Large kernel needs
+- Bitcoin (11300) - Large kernel needs
 
-· SHA512 (1700) - 512-bit operations
+- SHA512 (1700) - 512-bit operations
 
 Recommended for Mobile
 
-· MD4/MD5 (800-1200 MH/s)
+- MD4/MD5 (800-1200 MH/s)
 
-· SHA1 (200-400 MH/s)
+- SHA1 (200-400 MH/s)
 
-· SHA256 (80-150 MH/s)
+- SHA256 (80-150 MH/s)
 
-· Dictionary attacks
+- Dictionary attacks
 
-· Educational use
+- Educational use
 
 ---
 
@@ -184,18 +196,18 @@ Real-world: 9-character password cracked in 90 seconds at 694 MH/s
 
 ---
 
-🎉 Done
+## 🎉 Done
 
-Your Android device is now ready for hashcat! Perfect for:
+Your Android device is now ready for **hashcat!** Perfect for:
 
-· Educational password security
+- Educational password security
 
-· Portable penetration testing
+- Portable penetration testing
 
-· On-the-go hash verification
+- On-the-go hash verification
 
-· Security research and learning
+- Security research and learning
 
 ---
 
-Tested on POCO X6 Pro • Android 15 • Termux 0.119.0
+Tested on **POCO X6 Pro** • **Android 16** • **Termux 0.119.0-beta.3**

--- a/docs/credits.txt
+++ b/docs/credits.txt
@@ -35,6 +35,11 @@ Gabriele "matrix" Gristina <matrix@hashcat.net> (@gm4tr1x)
 * Benchmarks initial code base
 * hashcat-toolchain docker
 
+Ahmed "ahmed-alnassif" Al-Nassif <mr.ahmed.nassif@gmail.com> (@ahmed-alnassif)
+
+* Android port with Termux support
+* Android build documentation and setup guide
+
 Jean-Christophe "Fist0urs" Delaunay <jean-christophe.delaunay@synacktiv.com> (@Fist0urs)
 
 * Kerberos TGS Rep enctype 23 kernel module

